daemon/api_sideload_n_try.go

```diff
@@ -123,97 +153,156 @@ func sideloadOrTrySnap(c *Command, body io.ReadCloser, boundary string) Response
 		}
 		return trySnap(c.d.overlord.State(), form.Values["snap-path"][0], flags)
 	}
-	flags.RemoveSnapPath = true
 
+	flags.RemoveSnapPath = true
 	flags.Unaliased = isTrue(form, "unaliased")
 	flags.IgnoreRunning = isTrue(form, "ignore-running")
-	systemRestartImmediate := isTrue(form, "system-restart-immediate")
 
-	origPath, tempPath, errRsp := form.SnapFileNameAndTempPath()
+	sideloadFlags := sideloadFlags{
+		Flags:       flags,
+		dangerousOK: isTrue(form, "dangerous"),
+	}
+
+	snapFiles, errRsp := form.GetSnapFiles()
+	if errRsp != nil {
+		return errRsp
+	}
+
+	st := c.d.overlord.State()
+	st.Lock()
+	defer st.Unlock()
+
+	var chg *state.Change
+	if len(snapFiles) > 1 {
+		chg, errRsp = sideloadManySnaps(st, form, snapFiles, sideloadFlags, user)
+	} else {
+		chg, errRsp = sideloadSnap(st, form, snapFiles[0], sideloadFlags)
+	}
 	if errRsp != nil {
 		return errRsp
 	}
 
+	systemRestartImmediate := isTrue(form, "system-restart-immediate")
+	if systemRestartImmediate {
+		chg.Set("system-restart-immediate", true)
+	}
+
+	ensureStateSoon(st)
+
+	// the handoff is only done when the unlock succeeds (instead of panicking)
+	// but this is good enough
+	pathsToNotRemove = make([]string, len(snapFiles))
+	for i, snapFile := range snapFiles {
+		pathsToNotRemove[i] = snapFile.tmpPath
+	}
+
+	return AsyncResponse(nil, chg.ID())
+}
+
+func sideloadManySnaps(st *state.State, form *Form, snapFiles []*uploadedSnap, flags sideloadFlags, user *auth.UserState) (*state.Change, *apiError) {
+	sideInfos := make([]*snap.SideInfo, len(snapFiles))
+	names := make([]string, len(snapFiles))
+	tempPaths := make([]string, len(snapFiles))
+	origPaths := make([]string, len(snapFiles))
+
+	for i, snapFile := range snapFiles {
+		si, apiError := readSideInfo(st, snapFile.tmpPath, snapFile.filename, flags)
+		if apiError != nil {
+			return nil, apiError
+		}
+
+		sideInfos[i] = si
+		names[i] = si.RealName
+		tempPaths[i] = snapFile.tmpPath
+		origPaths[i] = snapFile.filename
+	}
+
+	var userID int
+	if user != nil {
+		userID = user.ID
+	}
+
+	tss, err := snapstateInstallPathMany(context.TODO(), st, sideInfos, tempPaths, userID, &flags.Flags)
+	if err != nil {
+		return nil, errToResponse(err, tempPaths, InternalError, "cannot install snap files: %v")
+	}
+
+	msg := fmt.Sprintf(i18n.G("Install snaps %q from files %q"), names, origPaths)
```

Comment:
%q here produces `["name1", "name2"...]`  which is not what we use. These should use instead %s and strutil.Quoted. You can see examples in api_snaps.go I think.